src/_zkapauthorizer/replicate.py

```diff
@@ -23,14 +23,77 @@
 ]
 
 from collections.abc import Awaitable
+from io import BytesIO
 from typing import BinaryIO, Callable
 
+import cbor2
+from attrs import frozen
 from twisted.python.lockfile import FilesystemLock
 
 from .config import REPLICA_RWCAP_BASENAME
 from .tahoe import Tahoe, attenuate_writecap
 
 
+@frozen
+class Change:
+    """
+    Represent an item in a replication event stream
+    """
+
+    sequence: int  # the sequence-number of this event
+    statement: str  # the SQL statement string
+
+
+@frozen
+class EventStream:
+    """
+    A series of database operations represented as `Change` instances.
+    """
+
+    changes: tuple[Change]
+
+    def highest_sequence(self):
+        """
+        :return int: the highest sequence number in this EventStream (or
+            None if there are no events)
+        """
+        if not self.changes:
+            return None
+        return max(change.sequence for change in self.changes)
+
+    # XXX or should we pass in a writable stream to use instead?
+    def to_bytes(self) -> BinaryIO:
+        """
+        :returns BinaryIO: a producer of bytes representing this EventStream.
+        """
+        return BytesIO(
+            cbor2.dumps(
+                {
+                    "events": tuple(
+                        (event.sequence, event.statement.encode("utf8"))
+                        for event in self.changes
+                    )
+                }
+            )
+        )
+
+    # XXX versioning? or do we handle that higher up?
```

Comment:
We could either put the version inside the stream or in metadata we find about the stream (the filename or metadata in the direntry).

It would probably be essentially trivial to put a `.v1` on the end of the filename and then declare versioning solved until we need a version 2.

Reply:
In magic-folders, the versioning debate ended with "we should put something in the Tahoe directory" and that something was an `@metadata` file that is JSON shaped like `{"version": 1}` ... (the `@` works out well there because `@<text>` is otherwise illegal in a magic-folder name _and_ because that actual JSON is small enough to fit in a LIT cap.

So we could do something similar here and declare the entire backup schmea as "v1" and put a `metadata.json` or something to that effect in the mutable-dir.

(...but what I was actually ref'ing in the above comment is that I saw `v1` in some other namings but I wasn't clear if those are meant to encapsulate "this way of writing to Tahoe" or something else).